[SymmetricMemory] introduce user-facing APIs empty() and rendezvous()

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139677
* #139529

Previously `SymmetricMemory` only had private pybind APIs:
```python
from torch.distributed._symmetric_memory import _SymmetricMemory
t = _SymmetricMemory.empty_strided_p2p(
    size=(64,),
    stride=(1,),
    dtype=torch.float32,
    device=device,
)
symm_mem_hdl = _SymmetricMemory.rendezvous(t, group_name=group.group_name)
```

This PR introduces user-facing APIs empty() and rendezvous():
```python
import torch.distributed._symmetric_memory as symm_mem
t = symm_mem.empty(64, device="cuda")
symm_mem_hdl = symm_mem.rendezvous(t, group_name=group.group_name)
```

Notable differences compared to the pybind APIs:
- `empty()` now resembles `torch.empty()`:
  - shape can either be an integer sequence or pack
  - no need to/can't specify stride anymore
  - device can either be `torch.device` or string
- `group_name` needs to be specified at rendezvous time as opposed to allocation time. See https://github.com/pytorch/pytorch/pull/139529 for the rationales. I feel the new semantic is superior, hence enforcing it in the public API.
  - Currently, the pybind API still support specifying `group_name` at rendezvous time. 

This PR does not change the behavior of the pybind APIs.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Commit:
[SymmetricMemory] initial user-facing API

[ghstack-poisoned]

Commit:
Update on "[SymmetricMemory] initial user-facing APIs"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

Commit:
Update on "[SymmetricMemory] initial user-facing APIs"




cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

Commit:
Update on "[SymmetricMemory] introduce user-facing APIs: empty() and rendezvous()"


Previously `SymmetricMemory` only had private pybind APIs:
```python
from torch.distributed._symmetric_memory import _SymmetricMemory
t = _SymmetricMemory.empty_strided_p2p(
    size=(64,),
    stride=(1,),
    dtype=torch.float32,
    device=device,
)
symm_mem_hdl = _SymmetricMemory.rendezvous(t, group_name=group.group_name)
```

This PR introduces user-facing APIs: empty() and rendezvous():
```python
import torch.distributed._symmetric_memory as symm_mem
t = symm_mem.empty(64, device="cuda")
symm_mem_hdl = symm_mem.rendezvous(t, group_name=group.group_name)
```

Notable differences compared to the pybind APIs:
- `empty()` now resembles `torch.empty()`:
  - shape can either be an integer sequence or pack
  - no need to/can't specify stride anymore
  - device can either be `torch.device` or string
- `group_name` needs to be specified at rendezvous time as opposed to allocation time. See https://github.com/pytorch/pytorch/pull/139529 for the rationales. I feel the new semantic is superior, hence enforcing it in the public API.
  - Currently, the pybind API still support specifying `group_name` at rendezvous time. 

This PR does not change the behavior of the pybind APIs.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

Commit:
Update on "[SymmetricMemory] introduce user-facing APIs empty() and rendezvous()"


Previously `SymmetricMemory` only had private pybind APIs:
```python
from torch.distributed._symmetric_memory import _SymmetricMemory
t = _SymmetricMemory.empty_strided_p2p(
    size=(64,),
    stride=(1,),
    dtype=torch.float32,
    device=device,
)
symm_mem_hdl = _SymmetricMemory.rendezvous(t, group_name=group.group_name)
```

This PR introduces user-facing APIs empty() and rendezvous():
```python
import torch.distributed._symmetric_memory as symm_mem
t = symm_mem.empty(64, device="cuda")
symm_mem_hdl = symm_mem.rendezvous(t, group_name=group.group_name)
```

Notable differences compared to the pybind APIs:
- `empty()` now resembles `torch.empty()`:
  - shape can either be an integer sequence or pack
  - no need to/can't specify stride anymore
  - device can either be `torch.device` or string
- `group_name` needs to be specified at rendezvous time as opposed to allocation time. See https://github.com/pytorch/pytorch/pull/139529 for the rationales. I feel the new semantic is superior, hence enforcing it in the public API.
  - Currently, the pybind API still support specifying `group_name` at rendezvous time. 

This PR does not change the behavior of the pybind APIs.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

lw commented on torch/distributed/_symmetric_memory/__init__.py:
```diff
@@ -110,6 +110,8 @@ def get_symm_mem_workspace(group_name: str, min_size: int) -> _SymmetricMemory:
         _SymmetricMemory: the symmetric memory workspace associated with the
         group.
     """
+    enable_symm_mem_for_group(group_name)
```
Ah this looks like a key change! Does it mean that now, if one calls the fused matmul+comms ops, they lazily initialize the SymmetricMemory on their own?

yifuwang replied:
Ah, right! I realized it's redundant to require users to call `enable_symm_mem_for_group(group_name)` when they are already explicitly using the group for symmetric memory. This indirectly addresses the tracing issue you encountered.

lw reviewed (approved):
I think the API looks good, I believe it boils down to the same "verbs" as before (it used to be two steps, it is still two steps) but they have become somewhat simplified (no more explicit strides, one can pass PG instances, ...) and they have become public, which makes it feel more "ok" to call them :)

The only question I have is how do you intend them to be imported in user code:
* The function names (`empty` and `rendezvous`) are quite common so I guess one won't import them directly (i.e., no `from torch.distributed._symmetric_memory import ...`).
* However, their fully qualified name is long, hence it's perhaps a bit unwieldy to invoke `torch.distributed._symmetric_memory.rendezvous` directly.
* Is the intended usage that people import the module with a shorthand (`import torch.distributed._symmetric_memory as symm_mem`) and then use that to invoke the functions (`symm_mem.empty(...)`)? This is what is done in the tests.

yifuwang commented:
> The only question I have is how do you intend them to be imported in user code:
> 
> * The function names (`empty` and `rendezvous`) are quite common so I guess one won't import them directly (i.e., no `from torch.distributed._symmetric_memory import ...`).
> * However, their fully qualified name is long, hence it's perhaps a bit unwieldy to invoke `torch.distributed._symmetric_memory.rendezvous` directly.
> * Is the intended usage that people import the module with a shorthand (`import torch.distributed._symmetric_memory as symm_mem`) and then use that to invoke the functions (`symm_mem.empty(...)`)? This is what is done in the tests.

The third bullet point is what I had in mind.

My thinking is that, since the API is already scoped at the module level, it might be redundant to further distinguish it at the function level. Here are two approaches we can recommend to users for working with the API:

```python
import torch.distributed._symmetric_memory as symm_mem

t = symm_mem.empty(...)
```

```python
from torch.distributed._symmetric_memory import empty as symm_mem_empty

t = symm_mem_empty(...)
```

This approach is also consistent with other `empty()` variants in distributed (e.g. [dtensor api](https://github.com/pytorch/pytorch/blob/b86b5349cbb2ac4f3c7907cf2ac0d14010841647/torch/distributed/tensor/_api.py#L1021)).

Let me know if you have further suggestions!

Commit:
Update on "[SymmetricMemory] introduce user-facing APIs empty() and rendezvous()"


Previously `SymmetricMemory` only had private pybind APIs:
```python
from torch.distributed._symmetric_memory import _SymmetricMemory
t = _SymmetricMemory.empty_strided_p2p(
    size=(64,),
    stride=(1,),
    dtype=torch.float32,
    device=device,
)
symm_mem_hdl = _SymmetricMemory.rendezvous(t, group_name=group.group_name)
```

This PR introduces user-facing APIs empty() and rendezvous():
```python
import torch.distributed._symmetric_memory as symm_mem
t = symm_mem.empty(64, device="cuda")
symm_mem_hdl = symm_mem.rendezvous(t, group_name=group.group_name)
```

Notable differences compared to the pybind APIs:
- `empty()` now resembles `torch.empty()`:
  - shape can either be an integer sequence or pack
  - no need to/can't specify stride anymore
  - device can either be `torch.device` or string
- `group_name` needs to be specified at rendezvous time as opposed to allocation time. See https://github.com/pytorch/pytorch/pull/139529 for the rationales. I feel the new semantic is superior, hence enforcing it in the public API.
  - Currently, the pybind API still support specifying `group_name` at rendezvous time. 

This PR does not change the behavior of the pybind APIs.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

Commit:
Update on "[SymmetricMemory] introduce user-facing APIs empty() and rendezvous()"


Previously `SymmetricMemory` only had private pybind APIs:
```python
from torch.distributed._symmetric_memory import _SymmetricMemory
t = _SymmetricMemory.empty_strided_p2p(
    size=(64,),
    stride=(1,),
    dtype=torch.float32,
    device=device,
)
symm_mem_hdl = _SymmetricMemory.rendezvous(t, group_name=group.group_name)
```

This PR introduces user-facing APIs empty() and rendezvous():
```python
import torch.distributed._symmetric_memory as symm_mem
t = symm_mem.empty(64, device="cuda")
symm_mem_hdl = symm_mem.rendezvous(t, group_name=group.group_name)
```

Notable differences compared to the pybind APIs:
- `empty()` now resembles `torch.empty()`:
  - shape can either be an integer sequence or pack
  - no need to/can't specify stride anymore
  - device can either be `torch.device` or string
- `group_name` needs to be specified at rendezvous time as opposed to allocation time. See https://github.com/pytorch/pytorch/pull/139529 for the rationales. I feel the new semantic is superior, hence enforcing it in the public API.
  - Currently, the pybind API still support specifying `group_name` at rendezvous time. 

This PR does not change the behavior of the pybind APIs.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

Commit:
Update on "[SymmetricMemory] introduce user-facing APIs empty() and rendezvous()"


Previously `SymmetricMemory` only had private pybind APIs:
```python
from torch.distributed._symmetric_memory import _SymmetricMemory
t = _SymmetricMemory.empty_strided_p2p(
    size=(64,),
    stride=(1,),
    dtype=torch.float32,
    device=device,
)
symm_mem_hdl = _SymmetricMemory.rendezvous(t, group_name=group.group_name)
```

This PR introduces user-facing APIs empty() and rendezvous():
```python
import torch.distributed._symmetric_memory as symm_mem
t = symm_mem.empty(64, device="cuda")
symm_mem_hdl = symm_mem.rendezvous(t, group_name=group.group_name)
```

Notable differences compared to the pybind APIs:
- `empty()` now resembles `torch.empty()`:
  - shape can either be an integer sequence or pack
  - no need to/can't specify stride anymore
  - device can either be `torch.device` or string
- `group_name` needs to be specified at rendezvous time as opposed to allocation time. See https://github.com/pytorch/pytorch/pull/139529 for the rationales. I feel the new semantic is superior, hence enforcing it in the public API.
  - Currently, the pybind API still support specifying `group_name` at rendezvous time. 

This PR does not change the behavior of the pybind APIs.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge
